microros_extensions/zephyr_serial_transport.c

```diff
@@ -45,17 +69,17 @@ size_t uxr_write_serial_data_platform(uxrSerialPlatform* platform, uint8_t* buf,
 }
 
 size_t uxr_read_serial_data_platform(uxrSerialPlatform* platform, uint8_t* buf, size_t len, int timeout, uint8_t* errcode){ 
+    size_t read = 0;
+    int spent_time = 0;
 
-    size_t index = 0;
-
-    int ret = 0;
-    while (index < len && ret == 0)
-    {
-        ret = uart_poll_in(platform->uart_dev, &buf[index]);
-        if (ret == 0) {
-            index++;
-        }
+    while(ring_buf_is_empty(&in_ringbuf) && spent_time < timeout){
+        k_sleep(K_MSEC(1));
+        spent_time++;
```

Comment:
Is it deterministic? I mean, is Zephyr able to ensure that when a task sleeps it won't be longer than the request time?

Reply:
When dealing with preemptive tasks and no higher priority task exists it is deterministic. 
If there is a higher prio task, it may take longer. By default micro-ROS works on the preemptive task priorities ranges, so cooperative behavior is not contemplated.

Reply:
Ok, is there a way to measure time in a deterministic manner?

Reply:
Using clocks: https://docs.zephyrproject.org/1.9.0/kernel/timing/clocks.html